### ISSUE TITLE
"unused methods" false-positives

### ISSUE THREAD

**Issue by kit-ty-kate** [LexiFi/dead_code_analyzer#107]
```
git clone https://github.com/kit-ty-kate/opam
cd opam
git checkout 2d04f927492c9b7fb368d358513ccd8a7d2a9d66
./configure
make
dune build --profile=release --root . @check && dead_code_analyzer --all _build/default/src/**/.*objs/ | grep dose4
```
shows:
```
<pwd>/_build/default/src/solver/dose4/util.mli:44: identity#add
<pwd>/_build/default/src/solver/dose4/util.mli:44: identity#inttovar
<pwd>/_build/default/src/solver/dose4/util.mli:44: identity#vartoint
```
However looking around at uses of `Util.identity` we find very quickly find at least one user of these methods (at least for `inttovar` and `vartoint`):
```
Depsolver.listcheck --> Diagnostic.diagnosis --> Diagnostic.result
```

**Comment by fantazio** [comment]
I tried to reproduce this but didn't get any unused method reported.
Here is what the output of `./configure` looks like :
<details>
  <summary>Click to expand/hide</summary>

```
checking for ocamlc... ocamlc
OCaml version is 5.4.1
OCaml library path is <path_to_clone>/_opam/lib/ocaml
checking for ocamlopt... ocamlopt
checking for ocamlc.opt... ocamlc.opt
checking for ocamlopt.opt... ocamlopt.opt
checking for ocaml... ocaml
checking for ocamldep... ocamldep
checking for ocamldep.opt... ocamldep.opt
checking for ocamlmktop... ocamlmktop
checking for ocamlmklib... ocamlmklib
checking for ocamldoc... ocamldoc
checking for ocamldoc.opt... ocamldoc.opt
checking for ocamlbuild... ocamlbuild
checking OCaml Sys.os_type... Unix
checking for gawk... no
checking for mawk... mawk
checking for compiler type... cc
checking for compiler architecture... amd64
checking for compiler system... linux
checking for compiler target... x86_64-pc-linux-gnu
checking for gcc... gcc -O2 -fno-strict-aliasing -fwrapv -fPIC  -pthread   -D_FILE_OFFSET_BITS=64
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether the compiler supports GNU C... yes
checking whether gcc -O2 -fno-strict-aliasing -fwrapv -fPIC  -pthread   -D_FILE_OFFSET_BITS=64 accepts -g... yes
checking for gcc -O2 -fno-strict-aliasing -fwrapv -fPIC  -pthread   -D_FILE_OFFSET_BITS=64 option to enable C11 features... none needed
checking for ocamlobjinfo... ocamlobjinfo
checking for ocamlfind... ocamlfind
checking for curl... curl
checking for dune... dune
checking for patch... patch
checking for bunzip2... bunzip2
checking for linking method... shared

checking for OCaml findlib package unix... found
checking for OCaml findlib package re 1.10.0 or later... found [unspecified]
checking for OCaml findlib package base64 3.1.0 or later... found 3.5.2
checking for OCaml findlib package ocamlgraph 1.8.8 or later... found 2.2.0
checking for OCaml findlib package cudf 0.7 or later... found 0.11~rc1
checking for OCaml findlib package dose3.common 6.1 or later... found 7.0.0
checking for OCaml findlib package dose3.algo 6.1 or later... found 7.0.0
checking for OCaml findlib package opam-file-format 2.1.4 or later... found 2.1.6
checking for OCaml findlib package spdx_licenses 1.4.0 or later... found 1.6.0
checking for OCaml findlib package opam-0install-cudf 0.5.0 or later... found 0.5.0
checking for OCaml findlib package jsonm 1.0.2 or later... found 1.0.2
checking for OCaml findlib package uutf 1.0.3 or later... found 1.0.4
checking for OCaml findlib package sha 1.13 or later... found v1.15.4
checking for OCaml findlib package swhid_core 0.1 or later... found [unspecified]
checking for OCaml findlib package patch 3.0.0 or later... found 3.1.2
checking for OCaml findlib package tar 3.3.0 or later... found v3.5.0
checking for OCaml findlib package checkseum 0.5.2 or later... found 0.5.3
checking for OCaml findlib package decompress 1.5.1 or later... found [unspecified]
checking for OCaml findlib package mccs 1.1+17 or later... found 1.1+19
checking for g++... g++
checking whether the compiler supports GNU C++... yes
checking whether g++ accepts -g... yes
checking for g++ option to enable C++11 features... none needed
checking whether g++ really is a C++ compiler... yes

configure: creating ./config.status
config.status: creating Makefile.config
config.status: creating src/ocaml-flags-configure.sexp
config.status: creating src/core/c-flags.sexp

Opam will be built WITH its default built-in solver

Executables will be installed in /usr/local/bin
Manual pages will be installed in /usr/local/share/man
```
</details>

I used the `dead_code_analyzer.1.3.0` in OCaml `4.14.4`, `5.2.1`, `5.4.1` and `5.5.1`.
The only notable differences with your instructions are :
- I created a local switch via `opam switch create .` after the `git checkout`;
- after each OCaml version update, I ran `make distclean` before `./configure`.


Which version of the `dead_code_analyzer` are you using ?
Which version of OCaml are you using ?
Did you get a line like ``*** WARNING: <int> files cannot be read. Run with `--verbose` for more details.`` at the top of the reports, after scanning ? It's usually a good first indicator of the cause of FP

**Comment by kit-ty-kate** [comment]
mmh, i'm not sure what could be different between our two setups.
I used dead_code_analyzer.1.3.0 with ocaml-base-compiler.5.5.1 in a global switch (but i don't think that matters).
I didn't get any warnings

**Comment by kit-ty-kate** [comment]
I don't think that matters either but this was on Linux/arm64

**Comment by fantazio** [comment]
I found the issue in my setup: `globstar` was off. By enabling it I was able to reproduce your result.

Because the class `identity` is exported with a class type (`projection`) as signature, none of its methods should be reported, regardless of their use, according to the [documentation](https://github.com/LexiFi/dead_code_analyzer/blob/master/docs/methods/code_constructs/CLASS_TYPE.md) .
False positives confirmed.

**Comment by kit-ty-kate** [comment]
ah yes my bad, i use zsh as my default shell and forgot `**` isn't a portable construct
